File: NeuralNetworks-java/src/io/github/marcelovca90/nn/model/Perceptron.java
package io.github.marcelovca90.nn.model;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

import io.github.marcelovca90.nn.data.DataSet;
import io.github.marcelovca90.nn.math.ActivationFunction;
import io.github.marcelovca90.nn.math.MathUtils;
import io.github.marcelovca90.nn.plot.PlotUtils;

public class Perceptron implements NeuralNetwork
{
    private double[][] x;
    private double[] d;
    private double n;
    private double[] w;
    private ActivationFunction g;
    private List<Double> plotDataX;
    private List<Double> plotDataY;

    public Perceptron(double n, ActivationFunction g)
    {
        this.n = n;
        this.g = g;
        this.plotDataX = new ArrayList<>();
        this.plotDataY = new ArrayList<>();
    }

    @Override
    public double[] train(DataSet dataSet)
    {
        x = dataSet.getSamples();
        d = dataSet.getLabels();
        w = new Random(42L).doubles(dataSet.getNumberOfFeatures()).toArray();

        int epoch = 0;
        boolean error;

        do
        {
            error = false;
            for (int i = 0; i < dataSet.getNumberOfSamples(); i++)
            {
                double v = MathUtils.dotProduct(x[i], w);
                double y = g.compute(v);
                if (Double.compare(y, d[i]) != 0)
                {
                    for (int j = 0; j < dataSet.getNumberOfFeatures(); j++)
                    {
                        w[j] += (n * (d[i] - y) * x[i][j]);
                    }
                    error = true;
                }
            }
            epoch++;
            System.err.printf("Epoch: %d\tWeights: %s\n", epoch, Arrays.toString(w));
            plotDataX.add((double) epoch);
            plotDataY.add(error ? 1.0 : 0.0);

        } while (error && epoch < 10000);

        return Arrays.copyOf(w, w.length);
    }

    @Override
    public double test(double[] weights, double[] sample)
    {
        double v = MathUtils.dotProduct(w, sample);
        double y = g.compute(v);
        return y;
    }

    @Override
    public double evaluate(double[] weights, DataSet dataSet)
    {
        x = dataSet.getSamples();
        d = dataSet.getLabels();
        long correct = 0;
        long total = dataSet.getNumberOfSamples();
        for (int i = 0; i < dataSet.getNumberOfSamples(); i++)
        {
            double y = test(weights, x[i]);
            if (Double.compare(y, d[i]) == 0)
            {
                correct++;
            }
        }
        double accuracy = (double) correct / (double) total;
        System.out.printf("Accuracy: %.2f%% (%d/%d)\n", 100.0 * accuracy, correct, total);
        return accuracy;
    }

    public void plotErrorPerEpoch()
    {
        PlotUtils.plot(
            plotDataX.stream().mapToDouble(Double::doubleValue).toArray(),
            "epoch",
            plotDataY.stream().mapToDouble(Double::doubleValue).toArray(),
            "error");
    }
}

File: NeuralNetworks-python/blueprint/estimators/perceptron.py
import numpy as np

from .estimator import Estimator


class Perceptron(Estimator):
  def __init__(self, learning_rate, activation_function, epochs):
    super().__init__()

    self.learning_rate = learning_rate
    self.activation_function = activation_function
    self.epochs = epochs

  def train(self, x, d):
    k = len(x)
    w = np.random.rand(len(x[0]))
    epoch = 0
    error = True

    print(f'Epoch: {epoch}\tWeights: {w}')

    self.plot_data_x.append(epoch)
    self.plot_data_y.append(1 if error else 0)

    while error and epoch < self.epochs:
      error = False

      for i in range(k):
        v = np.dot(np.transpose(w), x[i])
        y = self.activation_function(v)

        if y != d[i]:
          w = np.add(w, np.multiply(self.learning_rate * (d[i] - y), x[i]))
          error = True

      epoch = epoch + 1

      print(f'Epoch: {epoch}\tWeights: {w}')

      self.plot_data_x.append(epoch)
      self.plot_data_y.append(1 if error else 0)

    return w

  def predict(self, x, w):
    v = np.dot(np.transpose(w), x)
    y = self.activation_function(v)

    return y

  def evaluate(self, x, d, w):
    total = len(x)
    correct = 0

    for i in range(len(x)):
      y = self.predict(x[i], w)

      if y == d[i]:
        correct = correct + 1

    accuracy = correct / total

    print(f'Accuracy: {accuracy * 100:.5f}% ({accuracy:.5f})')

    return accuracy
